tools/wptrunner/wptrunner/executors/protocol.py

```diff
@@ -28,6 +28,18 @@ def __init__(self, executor, browser):
             assert not hasattr(self, name)
             setattr(self, name, cls(self))
 
+    def update(self, target, source):
+        if (type(target) is not dict
+                or type(source) is not dict):
+            return False
+        rv = True
+        for sk in source:
+            if sk not in target:
+                target[sk] = source[sk]
+            else:
+                rv = rv and self.update(target[sk], source[sk])
```

Comment:
This only works if there aren't any non-dict properties to update, right? So if we had `update({"foo":True}, {"foo": False})` we'd end up returning `False` and not doing the update. Related: if the update can fail, maybe we should work on a copy rather than modifying in place to avoid non-atomic updates.

Reply:
Yes, this only meant to update dicts. I updated the code a little bit to allow `update({"foo": True}, {"foo": True})` to succeed. Also added roll back so that if this function returns False, target will not be updated.